ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -508,26 +546,44 @@ static void trigger_sample_for_thread(
   ddog_Slice_i64 metric_values_slice,
   sample_type type
 ) {
-  VALUE thread_name = thread_name_for(thread);
-  bool have_thread_name = thread_name != Qnil;
-
-  int label_count = 1 + (have_thread_name ? 1 : 0);
-  ddog_Label labels[label_count];
+  int max_label_count =
+    1 + // thread id
+    1 + // thread name
+    2;  // local root span id and span id
+  ddog_Label labels[max_label_count];
+  int label_pos = 0;
+
+  labels[label_pos++] = (ddog_Label) {
+    .key = DDOG_CHARSLICE_C("thread id"),
+    .str = thread_context->thread_id_char_slice
+  };
 
-  labels[0] = (ddog_Label) {.key = DDOG_CHARSLICE_C("thread id"), .str = thread_context->thread_id_char_slice};
-  if (have_thread_name) {
-    labels[1] = (ddog_Label) {
+  VALUE thread_name = thread_name_for(thread);
+  if (thread_name != Qnil) {
+    labels[label_pos++] = (ddog_Label) {
       .key = DDOG_CHARSLICE_C("thread name"),
       .str = char_slice_from_ruby_string(thread_name)
     };
   }
 
+  struct trace_identifiers trace_identifiers_result = {.valid = false};
+  trace_identifiers_for(state, thread, &trace_identifiers_result);
+
+  if (trace_identifiers_result.valid) {
+    labels[label_pos++] = (ddog_Label) {.key = DDOG_CHARSLICE_C("local root span id"), .str = trace_identifiers_result.local_root_span_id};
+    labels[label_pos++] = (ddog_Label) {.key = DDOG_CHARSLICE_C("span id"), .str = trace_identifiers_result.span_id};
+  }
+
+  if (label_pos > max_label_count) {
+    rb_raise(rb_eRuntimeError, "BUG: Unexpected label_pos (%d) > max_label_count (%d)", label_pos, max_label_count);
+  }
```

Comment:
Can this ever happen?

Reply:
It looks like someone has to modify the function for this to happen.
Does it make sense to have one of those "conditional assertion that does not exist in the production release"?

Reply:
> Can this ever happen?
> It looks like someone has to modify the function for this to happen.

Yes, exactly! I've pushed bb95c1c737f237a94d2c76e1856d94a142e358ca to better explain why that's there so you and others don't have to "reverse engineer" my intent ;)

> Does it make sense to have one of those "conditional assertion that does not exist in the production release"?

In this particular case, I don't think so, although I have been thinking about some mechanism to do that. But before I explain why, I'm curious why you ask:

* Is it because you're concerned about performance? If so, since this is a comparison between a constant and a local that recently got used, I'd expect this to be really cheap. In fact, I wonder if the compiler can be smart enough to optimize it out, by looking at the flow of `label_pos` in the function and proving it never can get bigger than `max_label_count`.

* Is it because you're concerned about correctness? The profiler is correctly built to handle exceptions. When they happen, we log them out and then stop profiling. I think for these "this should never happen" situations, stopping profiling entirely seems like a reasonable thing to do to avoid any impact on the user application.

---

That said, here's why "In this particular case, I don't think so": Because I fear that contributors and customers may not realize such a mode exists.

In fact, there's a known fork of dd-trace-rb that changes exactly this function -- 
https://github.com/zendesk/dd-trace-rb/commit/192e5c88437987276b395a3ecb8fcdda32cfe89d -- and when that fork gets updated with this change, I wanted to make it easier to update this code without issues.

In its current approach, any incorrect change to this function would appear quite soon in a quite dramatic way (the profiler stops), which I think would allow us/customers to catch any bugs sooner.

---

Finally, I have been thinking about having some mode that we could use in CI/testing to enable extra checks, **especially** extra checks that may be expensive/break customer setups and we wouldn't want to have on.

A good example is compiler warnings -- we don't know which version of the C compiler our customers are using, so using `-Werror` to turn compiler warnings into errors on the release version of the gem is a bad idea -- suddenly a newer version of the compiler gets released and the profiler stops working because there's a new warning.

BUUUT we could enable this mode for just for CI/development, and it would allow us to catch new warnings as we introduce them.

So yeah, I haven't yet looked into such a setup, but suggestions welcome. (But in this case, even if such a setup was already in place, I would still not move this check to be done only on CI/dev mode, for the reasons I explained above).

Reply:
Thanks for the detailed response.

I see this check as a compiler shortcoming: the compiler should catch this. So my thought was "let's augment the compilation step then".
It seems like leaving the check as is is the best option today, so I'm happy with that.